src/aggregate/aggregate_request.c

```diff
@@ -1072,6 +1088,31 @@ int AREQ_Compile(AREQ *req, RedisModuleString **argv, int argc, QueryError *stat
     goto error;
   }
 
+  // Define if we need a depleter in the pipeline to get accurate total results
+  if (IsAggregate(req) && HasWithCount(req) && !IsCount(req)) {
+    PLN_ArrangeStep *arng = AGPLN_GetArrangeStep(AREQ_AGGPlan(req));
+    bool isLimited = (arng && arng->isLimited && arng->limit > 0);
+
+    if (req->protocol == 2){
+      // In RESP2, total_results is returned as the first element in the array,
+      // so we need to deplete the pipeline to count the total number of results.
+      // When sorting or grouping, the depleter is not needed since these
+      // steps already deplete the pipeline.
+      if (!HasSortBy(req) && !HasGroupBy(req)) {
+        AREQ_AddRequestFlags(req, QEXEC_F_HAS_DEPLETER);
+      }
+    } else if (req->protocol == 3) {
```

Comment:
Isn't it also necessary to deplete in resp3 in order to get the accurate total count?
(also in resp3 a depleter is not necessarily going to be used by default)